src/Workspaces/CSharp/Portable/LanguageServices/CSharpTypeInferenceService.TypeInferrer.cs

```diff
@@ -1950,6 +1951,121 @@ private IEnumerable<TypeInferenceInfo> InferTypeInVariableDeclarator(VariableDec
                 return types;
             }
 
+            private IEnumerable<TypeInferenceInfo> InferTypeInVariableComponentAssignment(
+                VariableComponentAssignmentSyntax variableComponentAssigment, ExpressionSyntax expression)
+            {
+                if (expression == variableComponentAssigment.Value)
+                {
+                    var variableComponent = variableComponentAssigment.VariableComponent;
+                    if (variableComponent.IsKind(SyntaxKind.TypedVariableComponent))
+                    {
+                        var typedVariable = (TypedVariableComponentSyntax)variableComponent;
+                        return GetTypes(typedVariable.Type);
+                    }
+                    else if (variableComponent.IsKind(SyntaxKind.ParenthesizedVariableComponent))
+                    {
+                        // We have something of the form:
+                        //   (int a, int b) = ...
+                        //
+                        // This is a deconstruction, and a decent deconstructable type we can infer here
+                        // is ValueTuple<int,int>.
+                        var parenthesizedVariable = (ParenthesizedVariableComponentSyntax)variableComponent;
+                        var tupleType = GetTupleType(parenthesizedVariable);
+
+                        if (tupleType != null)
+                        {
+                            return SpecializedCollections.SingletonEnumerable(new TypeInferenceInfo(tupleType));
+                        }
+                    }
+                }
+
+                return SpecializedCollections.EmptyEnumerable<TypeInferenceInfo>();
+            }
+
+            private ITypeSymbol GetTupleType(
+                ParenthesizedVariableComponentSyntax parenthesizedVariableComponent)
+            {
+                ImmutableArray<ITypeSymbol> elementTypes;
+                ImmutableArray<string> elementNames;
+
+                if (!TryGetTupleTypesAndNames(parenthesizedVariableComponent.Variables,
+                        out elementTypes, out elementNames))
+                {
+                    return null;
+                }
+
+                return Compilation.CreateTupleTypeSymbol(elementTypes, elementNames);
+            }
+
+            private bool TryGetTupleTypesAndNames(
+                SeparatedSyntaxList<VariableComponentSyntax> variables,
+                out ImmutableArray<ITypeSymbol> elementTypes,
+                out ImmutableArray<string> elementNames)
+            {
+                elementTypes = default(ImmutableArray<ITypeSymbol>);
+                elementNames = default(ImmutableArray<string>);
+
+                var elementTypesBuilder = ArrayBuilder<ITypeSymbol>.GetInstance();
+                var elementNamesBuilder = ArrayBuilder<string>.GetInstance();
+                try
```

Comment:
try/finally to avoid having to free the builders in the null case seems too convoluted. When I see the "try" I start thinking "okay, something unexpected might happen here" instead of "oh, maybe there's some cleanup code at the end", but maybe that's just me.